core/src/main/java/com/google/zxing/qrcode/encoder/Encoder.java

```diff
@@ -78,64 +78,87 @@ public static QRCode encode(String content,
                               Map<EncodeHintType,?> hints) throws WriterException {
 
-    // Determine what character encoding has been specified by the caller, if any
-    Charset encoding = DEFAULT_BYTE_MODE_ENCODING;
-    boolean hasEncodingHint = hints != null && hints.containsKey(EncodeHintType.CHARACTER_SET);
-    if (hasEncodingHint) {
-      encoding = Charset.forName(hints.get(EncodeHintType.CHARACTER_SET).toString());
-    }
-
-    // Pick an encoding mode appropriate for the content. Note that this will not attempt to use
-    // multiple modes / segments even if that were more efficient. Twould be nice.
-    Mode mode = chooseMode(content, encoding);
-
-    // This will store the header information, like mode and
-    // length, as well as "header" segments like an ECI segment.
-    BitArray headerBits = new BitArray();
-
-    // Append ECI segment if applicable
-    if (mode == Mode.BYTE && hasEncodingHint) {
-      CharacterSetECI eci = CharacterSetECI.getCharacterSetECI(encoding);
-      if (eci != null) {
-        appendECI(eci, headerBits);
-      }
-    }
-
-    // Append the FNC1 mode header for GS1 formatted data if applicable
-    boolean hasGS1FormatHint = hints != null && hints.containsKey(EncodeHintType.GS1_FORMAT);
-    if (hasGS1FormatHint && Boolean.parseBoolean(hints.get(EncodeHintType.GS1_FORMAT).toString())) {
-      // GS1 formatted codes are prefixed with a FNC1 in first position mode header
-      appendModeInfo(Mode.FNC1_FIRST_POSITION, headerBits);
-    }
-
-    // (With ECI in place,) Write the mode marker
-    appendModeInfo(mode, headerBits);
-
-    // Collect data within the main segment, separately, to count its size if needed. Don't add it to
-    // main payload yet.
-    BitArray dataBits = new BitArray();
-    appendBytes(content, mode, dataBits, encoding);
-
     Version version;
-    if (hints != null && hints.containsKey(EncodeHintType.QR_VERSION)) {
-      int versionNumber = Integer.parseInt(hints.get(EncodeHintType.QR_VERSION).toString());
-      version = Version.getVersionForNumber(versionNumber);
-      int bitsNeeded = calculateBitsNeeded(mode, headerBits, dataBits, version);
-      if (!willFit(bitsNeeded, version, ecLevel)) {
-        throw new WriterException("Data too big for requested version");
-      }
+    BitArray headerAndDataBits;
+    Mode mode;
+
+    boolean hasGS1FormatHint = hints != null && hints.containsKey(EncodeHintType.GS1_FORMAT) && Boolean.parseBoolean(hints.get(EncodeHintType.GS1_FORMAT).toString());
+    boolean hasCompactionHint = hints != null && hints.containsKey(EncodeHintType.QR_COMPACT) && Boolean.parseBoolean(hints.get(EncodeHintType.QR_COMPACT).toString());
+
+    if (hasCompactionHint) {
+        mode = Mode.BYTE;
+
+        MinimalEncoder.ResultNode rn = MinimalEncoder.encode(content,null,hasGS1FormatHint);
+        while (!willFit(rn.getSize(), rn.getVersion(ecLevel), ecLevel)) {
+            if (rn.getVersion(ecLevel).getVersionNumber() <= 26) {
```

Comment:
2-space indent everywhere

Reply:
Arghh, sorry. A pity that the style tool doesn't check that too.

Reply:
I just added / enforced checks for line length and indent, FWIW. It may flag some of this PR now

Reply:
I failed to figure out how I could update the branch from master so I changed pom.xml and checkstyle.xml temporarily the way you did and fixed/pushed the new issues on Encoder and MinimalEncoder that it found. I also fixed the indentation of Encoder.willFit() so that there shouldn't be a merge conflict later.